pkg/reporting/send.go

```diff
@@ -59,51 +82,54 @@ func (s eventSender[E]) senderLoop(ctx context.Context, logger *zap.Logger) {
 	}
 }
 
-func (s eventSender[E]) sendAllCurrentEvents(logger *zap.Logger) {
-	logger.Info("Pushing all available events")
+func (s eventSender[E]) sendAllCompletedBatches(logger *zap.Logger) {
+	logger.Info("Pushing all available event batches")
 
-	if s.queue.size() == 0 {
-		logger.Info("No events to push")
+	if len(s.queue.peekCompleted()) == 0 {
+		logger.Info("No event batches to push")
 		s.lastSendDuration = 0
 		s.metrics.lastSendDuration.WithLabelValues(s.client.Name).Set(1e-6) // small value, to indicate that nothing happened
 		return
 	}
 
-	total := 0
+	totalEvents := 0
+	totalBatches := 0
 	startTime := time.Now()
 
-	// while there's still events in the queue, send them
+	// while there's still batches of events in the queue, send them
 	//
-	// If events are being added to the queue faster than we can send them, this loop will not
-	// terminate. For the most part, that's ok: worst-case, we miss the collectorFinished
-	// notification, which isn't the end of the world. Any long-running call to this function will
-	// be reported by s.metrics.lastSendDuration as we go (provided the request timeout isn't too
-	// long).
+	// If batches are being added to the queue faster than we can send them, this loop will not
+	// terminate. For the most part, that's ok: worst-case, we miss that the parent context has
+	// expired, which isn't the end of the world (eventually the autoscaler-agent will just be
+	// force-killed). Any long-running call to this function will be reported by
+	// s.metrics.lastSendDuration as we go (provided the request timeout isn't too long), so we
+	// should get observability for it either way.
 	for {
-		if size := s.queue.size(); size != 0 {
-			logger.Info("Current queue size is non-zero", zap.Int("queueSize", size))
-		}
+		batches := s.queue.peekCompleted()
```

Comment:
question: It looks like we never use batches as a collection of batches, rather sticking with the first completed batch only. Should the queue API be changed to something like "peekLatestCompleted" and "peekCompletedCount"?

Reply:
Good point. Fixed in 6425e140142c2365ee3be7bf60957d9808bcf7da